lib/services/access_checker.go

```diff
@@ -321,24 +322,35 @@ func (a *accessChecker) CheckAccess(r AccessCheckable, state AccessState, matche
 func (a *accessChecker) GetKubeResources(cluster types.KubeCluster) (allowed, denied []types.KubernetesResource) {
 	if len(a.info.AllowedResourceIDs) > 0 {
 		for _, r := range a.info.AllowedResourceIDs {
-			if r.Kind == types.KindKubePod && r.Name == cluster.GetName() && r.ClusterName == a.localCluster {
+			if r.Name != cluster.GetName() || r.ClusterName != a.localCluster {
+				continue
+			}
+			switch {
+			case slices.Contains(types.KubernetesResourcesKinds, r.Kind):
 				splitted := strings.SplitN(r.SubResourceName, "/", 3)
 				// This condition should never happen since SubResourceName is validated
 				// but it's better to validate it.
 				if len(splitted) != 2 {
 					continue
 				}
+				// When the user is granted access through a Pod resource access request,
+				// he has no denied resources, which results in the denied being an empty slice.
 				allowed = append(allowed,
 					types.KubernetesResource{
 						Kind:      r.Kind,
 						Namespace: splitted[0],
 						Name:      splitted[1],
 					},
 				)
+			case r.Kind == types.KindKubernetesCluster:
+				// When a user has access to a Kubernetes cluster through Resource Access request,
+				// he has access to all resources in that cluster that he has access to through his roles.
+				// In that case, we append the allowed and denied resources from the roles.
+				roleAllowed, roleDenied := a.RoleSet.GetKubeResources(cluster)
+				allowed = append(allowed, roleAllowed...)
+				denied = append(denied, roleDenied...)
```

Comment:
Let's say a user creates a request for a specific pod in `cluster_a`, as well as all of `cluster_b`, in the same request, like this:
```sh
$ tsh request create --resource /teleport.example.com/pod/cluster_a/development/nginx-1 --resource /teleport.example.com/kube_cluster/cluster_b
```

and they have this `kube-access` role in their `search_as_roles`:
```yaml
kind: role
version: v6
metadata:
  name: kube-access
spec:
  allow:
    kubernetes_labels:
      '*':'*'
    kubernetes_resources:
      - kind: pod
        namespace: "*"
        name: "*"
    kubernetes_groups:
      - viewers
```

What I would expect is that the only pod in `cluster_a` they could access is `cluster_a/development/nginx-1` and e.g. trying to access `cluster_a/prod/app` would be denied. But I think this code as written would allow access to `cluster_a/prod/app`, since `CheckAccess` would allow access to everything under `cluster_a` and `matchKubernetesResource` would allow access to all pods in all namespaces due to the cluster request

I think this code block is confusing because it's trying to combine role access checks and resource request checks into a single set of allow/deny matchers that can be resolved in a single pass. If you just check that the roleset allows access first, and then check of a resource request match in a second pass, I bet it could be simplified

Reply:
discussed on slack, my example was wrong, I think I was confused because accessChecker.GetKubeResources did not really consider which pods were allowed by the roleset in all cases, just the ones allowed by the access requests. I didn't realize accessCheck.CheckAccess was also checking pod access based on the roleset. I think GetKubeResources is correct in the most recent commit